### DISCUSSION TITLE
Using Self-Deployed Certificate Secret w/ Teleport Cluster Chart

### DISCUSSION THREAD

**Discussion by mattmelgard** [gravitational/teleport#11101] [Q&A] [answered]
Hi there!

I'm trying to determine if there's a way to do this currently or not, or if this needs to be a feature request, but at the moment it seems the only documented way to deploy a HA Teleport Cluster in GCP mode is to [install cert-manager as a dependency](https://goteleport.com/docs/kubernetes-access/helm/guides/gcp/#step-4-install-and-configure-cert-manager) to provision the certificates necessary for TLS. We currently manage TLS certificates in our cluster by installing our own origin certificate from Cloudflare into the cluster itself as a TLS secret, but I don't see a way to define and use your own TLS secret on the upstream chart values. Is there a way to do this, am I missing something?

Thanks for any support!

**UPDATE:**
I see that this has already been asked [here](https://github.com/gravitational/teleport/discussions/9598) but it doesn't seem to have been resolved yet?

**Comment by webvictim** [comment, marked as answer]
There isn't currently an easy way to mount your own certs for Teleport to use with the `teleport-cluster` chart.

You can achieve what you'd like by doing the following:
- add your certificate as a TLS secret in Kubernetes
- install the chart using the `extraVolumes` and `extraVolumeMounts` values to mount your secret into the Teleport pod
- manually editing the `ConfigMap` (`kubectl -n <namespace> edit configmap/<release-name>`) to change the `proxy_service.https_keypairs` section to point to the path you defined:

```yaml
proxy_service:
  https_keypairs:
  - key_file: /path/to/key.pem
    cert_file: /path/to/cert.pem
```

You may also need to change the `chartMode` value to `custom` so that it doesn't overwrite your changed `ConfigMap`.

After this, restart the deployment and it should use your custom certificate.

**Reply by mattmelgard** [reply]
Hey @webvictim, I definitely considered doing that, but I feel that also disables a fair bit of the automation that is provided by the chart.

Since it seems like you're a maintainer of the upstream, is it possible that I could provide a simple contribution that would allow specifying a TLS cert/key secret in k8s and conditionally mount it and provide the location in the configmap of the chart? Similar to what's being done [here](https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-cluster/templates/config.yaml#L101) already?

**Reply by webvictim** [reply]
I'm working on getting unit tests merged for Helm at the moment, but my intention once that's done is actually to write this PR myself. It's coming up more and more and I think having first-party support for it is a good idea. I'm hoping I'll be able to get to it next week.

**Reply by mattmelgard** [reply]
Sounds good, I appreciate the transparency!
